src/main/java/org/opensearch/plugin/insights/core/listener/QueryInsightsListener.java

```diff
@@ -98,18 +108,61 @@ public void setEnableTopQueries(final MetricType metricType, final boolean isCur
         boolean isTopNFeatureCurrentlyDisabled = !queryInsightsService.isTopNFeatureEnabled();
 
         if (isTopNFeatureCurrentlyDisabled) {
-            super.setEnabled(false);
             if (!isTopNFeaturePreviouslyDisabled) {
-                queryInsightsService.checkAndStopQueryInsights();
+                checkAndStopQueryInsights();
             }
         } else {
-            super.setEnabled(true);
             if (isTopNFeaturePreviouslyDisabled) {
-                queryInsightsService.checkAndRestartQueryInsights();
+                checkAndRestartQueryInsights();
             }
         }
     }
 
+    /**
+     * Set search query metrics enabled to enable collection of search query categorization metrics
+     * @param searchQueryMetricsEnabled boolean flag
+     */
+    public void setSearchQueryMetricsEnabled(boolean searchQueryMetricsEnabled) {
+        boolean oldSearchQueryMetricsEnabled = queryInsightsService.isSearchQueryMetricsFeatureEnabled();
+        this.queryInsightsService.enableSearchQueryMetricsFeature(searchQueryMetricsEnabled);
+        if (searchQueryMetricsEnabled) {
+            if (!oldSearchQueryMetricsEnabled) {
+                checkAndRestartQueryInsights();
+            }
+        } else {
+            if (oldSearchQueryMetricsEnabled) {
+                checkAndStopQueryInsights();
+            }
+        }
+    }
```

Comment:
These logics are starting to get complicated. We should make these stop, starts idempotent and not rely on the old state

Reply:
refactored the logic